### DISCUSSION TITLE
Migrate from `underscore` to `lodash`?

### DISCUSSION THREAD

**Discussion by jmuchovej** [NYUCCL/psiTurk#542] [Ideas]
I'm curious if there's a reason why psiTurk continues to use `underscore` in place of `lodash`? (I definitely understand if it's "because legacy", but in my experience, `lodash` often more capable than `underscore`. 😕)

https://stackoverflow.com/a/13898916/2714651

**Comment by deargle** [comment]
I forgot we had a "discussions" area!

It's definitely a "because legacy" reason. psiTurk started in 2012!

**Comment by deargle** [comment]
Oh, and psiturk.js uses Backbone.js, which has underscore.js as (its only) hard dependency. https://backbonejs.org/

**Reply by jmuchovej** [reply]
Hmm… I see. I’m not super familiar with Backbone.js. What do you think it would take to replace it? 🤔

**Reply by deargle** [reply]
A huge factor driving design decisions is that a substantial psiturk userbase demographic is grad students who have never touched a command-line before, but whose lab leads ask them to figure out psiturk, or adapt a previous experiment. They struggle enough as it is without ever getting to the point of considering adding other external packages. Webpack or anything like it would likely be well outside their ability. 

Another very different demographic is power users who arrive here looking for ways to automate experiments. Someone like that could figure out Vue, or Angular, or React, or Next, or anything, but I'm hesitant to fully commit psiturk to any library like that, given the challenges of longterm maintainability. One thing that backbone has going for it is that it has worked since 2012! Psiturk_js isn't very complicated, nor perhaps need it be.

Maybe what we need is recipes for adapting to these other suites. And maybe we provide a few different automated builds of psiturk.js, to include an importable version.

**Reply by jmuchovej** [reply]
Hmm... I definitely see what you mean – I fall into both the grad student camp (and maybe a power-user, since many in the lab tend to come to me with questions 😅).

I understand "longterm maintainability" in general, but I'm not sure what you mean here. Could you elaborate? 

Also, while Backbone has been working since 2012, the last non-patch release was 3 years ago... I'm not sure if that's problematic for psiTurk, but there may be more useful packages now. ¯\_(ツ)_/¯ (Just a note – I haven't dug too deeply into `psiturk.js`, so this could be a uninformed opinion.)

Having a way to import `psiturk.js` that would allow for relatively straightforward use with <insert SSG/SPA framework> would be nice. (I haven't done this myself, but I _think_ I'd be willing to contribute to that.)

**Comment by jacob-lee** [comment]
Psiturk.js is perhaps one of the more underdeveloped parts of psiturk...

**Reply by deargle** [reply]
Indeed. Today, I wrestled with how to basically use the psiturk server without using psiturk_js, because I want to use a full Vue.js single page application (SPA), because I like shiny things. I concluded that it might be easiest to have exp.html forward the user to an entirely different website, where I would be running Vue, and which would receive the workerid etc. The only thing *required* on psiturk's end is that at some point, I post the user data back to `/sync/<uniqueid>`, and that I redirect to the `/complete`  route. What I lose from not using psiturk_js is basically not much. 

* Psiturk_js's use of Backbone is an alternative way to do an SPA via the `currentView` calls. 
* It also has event listeners bound to to page resizes and focus events
* It maintains a datastore for each participant, and has `saveData`, which calls a psiturk-server api for saving to the database
* It has some convenience functions for saving trial data with current timestamps

**Reply by jmuchovej** [reply]
I too like shiny things. Though my shiny thing has been `Next.js`. 😅 I’ve toyed with the idea of using it in place of Flask and friends, especially since it would allow for more easily adding `npm` packages. (Rather than hoping they have distributable JS files.)

It would also allow folks to write things like consent statements/whatnot in Markdown (rather than HTML). Though moving away from Python could make custom API routes a bit more challenging. 😕

**Reply by deargle** [reply]
For anyone following along and curious, I figured out a way to build a SPA with vue-cli, and to have it served via a Flask blueprint, which I register in psiturk's custom.py. The SPA will be the entirety of the `task.js` / exp.html page. It calls back to Flask api routes to mark when the participant has completed the experiment, etc. Flask is still responsible for serving the ad and directly interacting with the psiturk database. lmk if you're wanting to do something similar, and I can share more.
